rosbag2_storage/include/rosbag2_storage/qos.hpp

```diff
@@ -0,0 +1,164 @@
+
+namespace rosbag2_storage
+{
+/// Simple wrapper around rclcpp::QoS to provide a default constructor for YAML deserialization.
+class ROSBAG2_STORAGE_PUBLIC Rosbag2QoS : public rclcpp::QoS
+{
+public:
+  Rosbag2QoS()
+  : rclcpp::QoS(rmw_qos_profile_default.depth) {}
+
+  explicit Rosbag2QoS(const rclcpp::QoS & value)
+  : rclcpp::QoS(value) {}
+
+  Rosbag2QoS & default_history()
+  {
+    keep_last(rmw_qos_profile_default.depth);
+    return *this;
+  }
+
+  // Create an adaptive QoS profile to use for subscribing to a set of offers from publishers.
+  /**
+    * - Uses rosbag2_storage defaults for History since they do not affect compatibility.
+    * - Adapts Durability and Reliability, falling back to the least strict publisher when there
+    * is a mixed offer. This behavior errs on the side of forming connections with all publishers.
+    * - Does not specify Lifespan, Deadline, or Liveliness to be maximally compatible, because
+    * these policies do not affect message delivery.
+    */
+  static Rosbag2QoS adapt_request_to_offers(
+    const std::string & topic_name,
+    const std::vector<rclcpp::TopicEndpointInfo> & endpoints);
+
+  // Create a QoS profile to offer for playback.
+  /**
+    * This logic exists because rosbag2 does not record on a per-publisher basis, so we try to
+    * get as close as possible to the original system's behavior, given a single publisher.
+    * If all profiles are the same (excepting History & Lifespan, which are purely local),
+    * that exact value is returned.
+    * Otherwise, fall back to the rosbag2 default and emit a warning.
+    */
+  static Rosbag2QoS adapt_offer_to_recorded_offers(
+    const std::string & topic_name,
+    const std::vector<Rosbag2QoS> & profiles);
+};
+
+std::vector<rosbag2_storage::Rosbag2QoS> from_rclcpp_qos_vector(std::vector<rclcpp::QoS> in);
+std::string serialize_rclcpp_qos_vector(std::vector<rclcpp::QoS> in);
+std::vector<rclcpp::QoS> to_rclcpp_qos_vector(std::vector<rosbag2_storage::Rosbag2QoS> in);
+std::vector<rclcpp::QoS> to_rclcpp_qos_vector(YAML::Node in, int version);
+std::vector<rclcpp::QoS> to_rclcpp_qos_vector(std::string serielized, int version);
+}  // namespace rosbag2_storage
+
+namespace YAML
+{
+
+/// Pass metadata version to the sub-structs of BagMetadata for deserializing.
+/**
+  * Encoding should always use the current metadata version, so it does not need this value.
+  * We cannot extend the YAML::Node class to include this, so we must call it
+  * as a function with the node as an argument.
+  */
+
+template<typename T>
+T decode_for_version(const Node & node, int version)
+{
+  static_assert(
+    std::is_default_constructible<T>::value,
+    "Type passed to decode_for_version that has is not default constructible.");
+  if (!node.IsDefined()) {
+    throw TypedBadConversion<T>(node.Mark());
+  }
+  T value{};
+  if (convert<T>::decode(node, value, version)) {
+    return value;
+  }
+  throw TypedBadConversion<T>(node.Mark());
+}
+
```

Comment:
I think I did it because if you `#include "rosbag2_storage/qos.hpp` without including also `yaml.hpp` then you can't use it. And as said some days ago we can't include `yaml.hpp` in `qos.hpp` due to circular dependency. If I move it back, I foresee I will need to include `yaml.hpp` instead of `qos.hpp` in some places. Let me know what you prefer, nevertheless I will give it a try again and report back :)

Reply:
If I keep the old code:
```cpp
qos
  .keep_last(node["depth"].as<int>())
  .history(history)
```
I get the following error:
```shell
FAIL: test_info.TestRos2BagInfo.test_info_with_topic_name_option
--------------------------------------------------------------------------------------------------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/roxy/ros2_rolling/src/ros2/rosbag2/ros2bag/test/test_info.py", line 68, in test_info_with_topic_name_option
    assert normalize_lineseps(bag_command.output) == EXPECTED_OUTPUT, \
AssertionError: ros2bag CLI did not produce the expected output
assert '[WARN] [1698...ts\n/rosout\n' == '/parameter_events\n/rosout\n'
  + [WARN] [1698932598.894371851] [rclcpp]: A zero depth with KEEP_LAST doesn't make sense; no data could be stored.This will be interpreted as SYSTEM_DEFAULT
```
that looking at `rclcpp` source code, happens when you have a `depth=0`. See https://github.com/ros2/rclcpp/pull/2048. 

My switch/case was to set `depth` only when meaningful, i.e. when using `keep_last` strategy.

Reply:
I checked, but I noticed I need `decode_for_version` in `qos.hpp` and I can't introduce a circular dependency.